src/poliastro/twobody/events.py

```diff
@@ -25,8 +30,84 @@ def terminal(self):
     def last_t(self):
         return self._last_t * u.s
 
-    def __call__(self, t, u, k):
+    def __call__(self, t, u_, k):
         self._last_t = t
-        H = norm(u[:3])
+        H = norm(u_[:3])
         # SciPy will search for H - R = 0
         return H - self._R
+
+
+class PenumbraEvent:
+    """Detect whether a satellite is in penumbra or not.
+
+    Parameters
+    ----------
+    r_sun: ~astropy.units.Quantity
+        Position vector of the Sun with respect to the Earth.
+    R: ~astropy.units.Quantity
+        Radius of the attractor.
+
+    """
+
+    def __init__(self, r_sun, r_s, r_p, terminal=False):
+        self._r_sun = r_sun
+        self._r_s = r_s
+        self._r_p = r_p
+        self._last_t = None
+        self._terminal = terminal
+
+    @property
+    def terminal(self):
+        return self._terminal
+
+    @property
+    def last_t(self):
+        return self._last_t * u.s
+
+    def __call__(self, t, u_, k):
+        self._last_t = t
+        r_sat = (u_[:3] << u.km).value
+        r_sun = self._r_sun.to(u.km).value
```

Comment:
This might be a critical point. Vallado only provides the algorithm for a particular instant of time. But in this logic, we are propagating the orbit, which moves together with its attractor. I would like to ping @astrojuanlu here for a further discussion on this.

From my side, I will try to implement validation cases related to eclipse detection so we can be sure if these is properly implemented.